scripts/pre_commit_linter.py

```diff
@@ -143,6 +143,12 @@
             'extensions/classifiers/',
             'extensions/objects/',
             'extensions/value_generators/')
+    },
+    {
+        'regexp': r"([^ \t\r\n])[ \t]+$",
```

Comment:
This is interesting because it seems to be relating to a line and not a file. If these are per-line checks now, then let's rename the constant to BAD_LINE_PATTERNS_HTML_REGEXP for clarity.

Can you also explain why you need the capture group before `[ \t]+`?

Reply:
Okay, got it. Re the regex, this regex captures the trailing whitespaces after the non-whitespace characters while also preserving whitespace-only lines.

Reply:
Why do we want to preserve whitespace-only lines (in fact, why do we have them at all)?

Reply:
There were already a number of whitespace-only lines present in the files. (Maybe for clarity purposes). Therefore I did not remove them.

Reply:
I think they can be removed. Can you point me to one? I can tell you for certain then.

Reply:
[Line 9](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/pages/admin/roles_tab/roles_tab_directive.html#L9), [Line 27](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/pages/admin/roles_tab/roles_tab_directive.html#L27), [Line 42](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/pages/admin/roles_tab/roles_tab_directive.html#L42), [Line 49](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/pages/admin/roles_tab/roles_tab_directive.html#L49), [Line 67](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/pages/admin/roles_tab/roles_tab_directive.html#L67), [Line 77](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/pages/admin/roles_tab/roles_tab_directive.html#L77) and [Line 90](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/pages/admin/roles_tab/roles_tab_directive.html#L90) in `roles_tab_directive.html`.

Reply:
AFAICT (by pressing Ctrl+A), most of those lines are truly empty. Those lines can be left as-is, since they don't end with a trailing space in any case.

However lines 85 and 90 do have spaces. Those spaces should be removed.

Reply:
Okay. I'll have to change the RegEx then to `[ \t]+$`. It checks for those spaces as well. (I have tried it locally).